rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+# RESTful Service Specifications
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Goals
+
+ - Uniformity between data-centric APIs and remote procedure APIs
+ - Use of pre-existing standards/conventions when possible
+ - Minimization of bike-shedding
+
+Standards/Conventions used as inspiration:
+ - [Whitehouse API Standards](https://github.com/WhiteHouse/api-standards)
+  - RESTful URLs
+ - [Spring Data REST](http://docs.spring.io/spring-data/rest/docs/current/reference/html/)
+  - paging conventions
+  - uses [JSON HAL](https://tools.ietf.org/html/draft-kelly-json-hal-08)
+ - [JSend](http://labs.omniti.com/labs/jsend)
+  - API response conventions
+
+## URIs
+
+Version granularity is up to implementation. However, the base URI for an
+entity must include a version.
+
+The base URI for a given resource type is based on the *singular* name of the
+resource. Example:
+
+    https://example.com/rest/v1/user
+
+## Operations Against a Resource Type
+
+The following operations may be supported:
+
+| HTTP Operation | Description            | Query Parameters              | Request Body | Response Body  |
+|----------------|------------------------|-------------------------------|--------------|----------------|
+| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
+| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
+| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
+| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
+| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
+
+Endpoints may not support all methods (i.e. a service may provide read-only
+access to a resource, or a service may not support certain operations for
+performance or other considerations; for example, `GET` may not be supported
+for `user` if listing all users is disallowed for performance reasons.
+
+## Operations Against Specific Resources
+Specific resources will be addressed through the base URI plus the identifier
+for that resource.
+
+    /user/{id}
```

Comment:
Something like this might make more sense:

 GET path/id -> retrieve by id, only for simple identities, won't work for
composite id
 POST path/search?<query> -> search operation

This is following the convention that GET without a verb means retrieve
entity with id, and POST with a verb means perform an operation.

On Fri, Jun 24, 2016 at 1:41 PM, dewayneHat <notifications@github.com>
wrote:

> In rest_specification/restful_service_specification.md
> <https://github.com/lightblue-platform/lightblue-docs-specs/pull/43#discussion_r68451929>
> :
>
> > +| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
> > +| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
> > +| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
> > +| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
> > +| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
> > +
> > +Endpoints may not support all methods (i.e. a service may provide read-only
> > +access to a resource, or a service may not support certain operations for
> > +performance or other considerations; for example, `GET` may not be supported
> > +for `user` if listing all users is disallowed for performance reasons.
> > +
> > +## Operations Against Specific Resources
> > +Specific resources will be addressed through the base URI plus the identifier
> > +for that resource.
> > +
> > +    /user/{id}
>
> /user/search?login=foo
> +1 for query params when filtering content.
>
> To echo @BVulaj <https://github.com/BVulaj>
> GET path/{id} makes identification of the resource more uniform and self
> descriptive to the consumer with little knowledge of the server code.
>
> GET path?x=foo&y=bar&z=car. For *every* resource that is navigable via
> query params *only*, the consumer has to determine for each combination
> of fields: Am I specifying search criteria, identifying a single resource,
> or multiple resources? Can I PUT the resource(s) I just received back to
> that same endpoint? We would need really strong documentation in order for
> a consumer to use the API properly.
>
> Where as GET /users/id and GET /users/search?x=foo&y=bar&z=car tell me
> more about the action I'm asking the server to perform. I know intuitively,
> that I can't PUT back on /users/search?... results. It doesn't make sense
> to try, and if I do, the server will save me and I should get a 404. Where
> as the PUT /users?..... might actually try to make that happen, maybe
> with unexpected results if I pass the wrong combination of fields.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-docs-specs/pull/43/files/19a07e2017455997f1b9c4bdb15edb6e3cceb8ce#r68451929>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADgDDV0TRZPbqnp8GI6118QILg4VrHztks5qPDL-gaJpZM4IopeE>
> .
>